File: 3731-find-missing-elements/3731-find-missing-elements.c
#include <stdlib.h>
#include <stdbool.h>

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* findMissingElements(int* nums, int numsSize, int* returnSize) {
    if (numsSize == 0) {
        *returnSize = 0;
        return NULL;
    }

    // Finding min and max in one pass
    int min = nums[0];
    int max = nums[0];
    for (int i = 1; i < numsSize; i++) {
        if (nums[i] < min) min = nums[i];
        if (nums[i] > max) max = nums[i];
    }

    int range = max - min + 1;
    
    // Allocate primitive boolean tracking buffer map
    bool* present = (bool*)calloc(range, sizeof(bool));
    for (int i = 0; i < numsSize; i++) {
        present[nums[i] - min] = true;
    }

    // Pre-calculate upper bound size needed for missing items to allocate exactly
    int maxPossibleMissing = range > 2 ? range - 2 : 0;
    int* tempResult = (int*)malloc(maxPossibleMissing * sizeof(int));
    int missingCount = 0;

    // Collecting missing elements in sorted order
    for (int i = 1; i < range - 1; i++) {
        if (!present[i]) {
            tempResult[missingCount++] = i + min;
        }
    }

    // Shrink allocation down to the exact match size found
    *returnSize = missingCount;
    int* result = (int*)malloc(missingCount * sizeof(int));
    for (int i = 0; i < missingCount; i++) {
        result[i] = tempResult[i];
    }

    free(present);
    free(tempResult);
    return result;
}

File: 3731-find-missing-elements/3731-find-missing-elements.py
class Solution:
    def findMissingElements(self, nums: list[int]) -> list[int]:
        if not nums:
            return []
            
        # Finding min and max in one pass
        minimum = nums[0]
        maximum = nums[0]
        for num in nums[1:]:
            if num < minimum: minimum = num
            if num > maximum: maximum = num
            
        # Using a flat list for O(1) direct offset index tracking
        range_len = maximum - minimum + 1
        present = [False] * range_len
        for num in nums:
            present[num - minimum] = True
            
        # Collecting missing elements in sorted order
        result = []
        for i in range(1, range_len - 1):
            if not present[i]:
                result.append(i + minimum)
                
        return result
